server/sonar-server/src/main/java/org/sonar/server/computation/taskprocessor/CeWorkerFactory.java

```diff
@@ -0,0 +1,34 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.taskprocessor;
+
+import java.util.List;
+
+public interface CeWorkerFactory {
```

Comment:
actually, the use of the Listener pattern is useless.

It's point seems to be to workaround the constraint of order of instantiation between `ClusterClient`, `CeWorkerFactory` and `CeProcessingSchedulerImpl` (where `CeWorkerFactory` is called) but it totally fails at that. 

The dependency still exists:
if `ClusterClient` is instantiated after `CeProcessingSchedulerImpl`, the listener will be added after all the calls to `CeWorkerFactory#create` which makes the listener useless.
It seems you are aware of this limitation as you added the following to the constructor of `ClusterClient`:
```
this.workerUUIDS.put(hzInstance.getLocalEndpoint().getUuid(), ceWorkerFactory.getWorkerUUIDs());
```

As I pointed during our discussion last week, it will be way more reliable and explicite to actually code this dependency in SQ's code:

1. add method `initialize` to `ClusterClient` (or whatever more appropriate name) which implementation is to call on WorkerFactory to get the current set of worker uuids and share it across the cluster
2. add a call to this method after this line https://github.com/SonarSource/sonarqube/blob/master/server/sonar-server/src/main/java/org/sonar/server/computation/queue/CeQueueInitializer.java#L59

This approach has the benefit of not sharing the uuid of workers before they are actually started. A failure can occur between the time `CeProcessingSchedulerImpl` is instantiated and `CeQueueInitializer` runs.